tools/tinygoize/main.go

```diff
@@ -2,81 +2,315 @@
-			if err != nil {
-				log.Fatalf("parsing\n%v\n:%v", string(b), err)
-			}
-		done:
-			for _, cg := range f.Comments {
-				for _, c := range cg.List {
-					if !strings.HasPrefix(c.Text, goBuild) {
-						continue
+		fset := token.NewFileSet() // positions are relative to fset
+		f, err := parser.ParseFile(fset, file, string(b), parser.ParseComments|parser.SkipObjectResolution)
+		if err != nil {
+			log.Fatalf("parsing\n%v\n:%v", string(b), err)
+		}
+
+		goBuildPresent := false
+
+	done:
+		// modify existing //go:build line
+		for _, cg := range f.Comments {
+			for _, c := range cg.List {
+				if !strings.HasPrefix(c.Text, goBuild) {
+					continue
+				}
+				goBuildPresent = true
+
+				contains := strings.Contains(c.Text, constraint)
+
+				if (builds && !contains) || (!builds && contains) {
+					log.Printf("Skipped, constraint up-to-date: %s\n", file)
+					continue nextFile
+				}
+
+				if builds {
+					re := regexp.MustCompile(`\(?\s*!tinygo\s+\|\|\s+tinygo.enable\s*\)?(\s+\&\&)?`)
+					c.Text = re.ReplaceAllString(c.Text, "")
+					log.Printf("Stripping build constraint %v\n", file)
+
+					// handle potentially now-empty build constraint
+					re = regexp.MustCompile(`^\s*//go:build\s*$`)
+					if re.MatchString(c.Text) {
+						c.Text = "//"
 					}
-					c.Text = goBuild + "!tinygo && (" + c.Text[len(goBuild):] + ")"
-					break done
+				} else {
+					c.Text = goBuild + "(" + constraint + ") && (" + c.Text[len(goBuild):] + ")"
 				}
+				break done
 			}
-			// Complete source file.
-			var buf bytes.Buffer
-			if err = p.Fprint(&buf, fset, f); err != nil {
-				log.Fatalf("Printing:%v", err)
+		}
+
+		if !builds && !goBuildPresent {
+			// no //go:build line found: insert one
+			var cg ast.CommentGroup
+			cg.List = append(cg.List, &ast.Comment{Text: goBuild + constraint})
+
+			if len(f.Comments) > 0 {
+				// insert //go:build after first comment
+				// group, assumed copyright. Doesn't seem
+				// quite right but seems to work.
+				cg.List[0].Slash = f.Comments[0].List[0].Slash + 1
+				f.Comments = append([]*ast.CommentGroup{f.Comments[0], &cg}, f.Comments[1:]...)
+			} else {
+				// prepend //go:build
+				f.Comments = append([]*ast.CommentGroup{&cg}, f.Comments...)
 			}
-			if err := os.WriteFile(file, buf.Bytes(), 0o644); err != nil {
-				log.Fatal(err)
+		}
+
+		// Complete source file.
+		var buf bytes.Buffer
+		if err = p.Fprint(&buf, fset, f); err != nil {
+			log.Fatalf("Printing:%v", err)
+		}
+		if err := os.WriteFile(file, buf.Bytes(), 0o644); err != nil {
+			log.Fatal(err)
+		}
+	}
+	return
+}
+
+func writeMarkdown(file *os.File, pathMD *string, tinygo *string, status BuildStatus) (err error) {
+
+	fmt.Fprintf(file, "---\n\n")
+	fmt.Fprintf(file, "DO NOT EDIT.\n\n")
+	fmt.Fprintf(file, "Generated via `go run tools/tinygoize/main.go`\n\n")
+	fmt.Fprintf(file, "%v\n\n", tinygoVersion(tinygo))
+	fmt.Fprintf(file, "---\n\n")
```

Comment:
Is there a reason you separate this over multiple `Fprintf`?
Why not define sth like [this](https://github.com/u-root/u-root/blob/1a80242f1fee21955d85c1e892c78e20a31ddf5f/cmds/exp/bzimage/bzimage.go#L40) with a format specifier and then do that? imo this would make it more readable

Reply:
I did one line down from this. Here I didn't because of the markdown backticks in :

Generated via `go run tools/tinygoize/main.go`\n\n

The backticks conflict with the golang string-literal delimiter and (apparently!) no way to escape them.